modules/packages/Channel.chpl

```diff
@@ -0,0 +1,605 @@
+            return (front == nil);
+        }
+
+        /* Pop the first value from the queue */
+        proc dequeue() : unmanaged Waiter(eltType) {
+            var waiter : unmanaged Waiter(eltType)?;
+            if front == back {
+                waiter = front;
+                front = nil;
+                back = nil;
+            }
+            else {
+                waiter = front;
+                front = front!.next;
+                front!.prev = nil;
+                waiter!.next = nil;
+            }
+            return waiter!;
+        }
+
+        /* Remove the specified entry from the queue */
+        proc dequeue(waiter : unmanaged Waiter(eltType)) {
+            if waiter == front {
+                dequeue();
+            }
+            else if waiter == back {
+                back = back!.prev;
+                back!.next = nil;
+            }
+            else if waiter.prev != nil && waiter.next != nil {
+                waiter.prev!.next = waiter.next;
+                waiter.next!.prev = waiter.prev;
+            }
+
+        }
+    }
+
+    record channel {
+        type eltType;
+        var channelObj : shared chan(eltType);
+
+        /*
+        Initialize a channel
+
+        :arg elt: The element type used for sending and receiving
+        :type elt: `type`
+
+        :arg size: Specify the maximum capacity for the channel ``bufferSize``.
+        :type size: `int`
+        */
+        proc init(type eltType, size = 0) {
+            this.eltType = eltType;
+            channelObj = new shared chan(eltType, size);
+        }
+
+        proc lock() {
+            this.channelObj.lock();
+        }
+
+        proc unlock() {
+            this.channelObj.unlock();
+        }
+
+        /*
+        Receive the first value in the channel buffer. It will suspend the
+        calling task, until data is sent to the channel. If the channel is
+        closed and the buffer is empty, it will return `false` indicating that
+        the receive operation was not successful.
+
+        :arg val: Storage for the received value.
+        :type val: `eltType`
+
+        :return: `true` if the receive was successful, else `false`.
+        :rtype: `bool`
+        */
+
+        proc recv(out val : eltType) : bool {
+            return recv(val, true);
+        }
+
+        pragma "no doc"
+        proc recv(out val : eltType, blocking : bool) : bool {
+            return this.channelObj.recv(val, blocking);
+        }
+
+        /*
+        Send a value to the channel buffer. If the buffer is at maximum
+        capacity it will suspend the waiting task, until there is space in the
+        buffer or a receiving task awakes it. If a channel is closed no more
+        data can be sent to it.
+
+        :arg val: Data to be sent to the channel
+        :type val: `eltType`
+
+        :throws ChannelError: If ``send`` is called on a closed channel.
+        */
+
+        proc send(in val : eltType) throws {
```

Comment:
The `in` intent here has similar implications as above. If the data is large we're copying it into the method. Which maybe the right thing given the concurrent nature of the interface here. But should there be a way to pass things by reference?

Reply:
Regarding the `recv(out)` part, see https://github.com/chapel-lang/chapel/pull/17954/files#r717006577 . However I don't think it should hold up this PR.

Regarding the `in` intent, I think that using `in` intent here is good because the channel will always (in normal operation) store the argument in the channel. With the `in` intent here, we can avoid copies completely if the argument is e.g. a value returned from a function. In contrast, if we used `const ref` here, the body of `send` would always have to copy in to the channel's buffer.